src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -0,0 +1,984 @@
+            self.qualif = self.qualif - Qualif::STATIC;
+            self.add(Qualif::NOT_CONST);
+
+            false
+        } else {
+            true
+        }
+    }
+
+    /// Assign the current qualification to the given destination.
+    fn assign(&mut self, dest: &Lvalue<'tcx>) {
+        let qualif = self.qualif;
+        let span = self.span;
+        let store = |slot: &mut Option<Qualif>| {
+            if slot.is_some() {
+                span_bug!(span, "multiple assignments to {:?}", dest);
+            }
+            *slot = Some(qualif);
+        };
+
+        // Only handle promotable temps in non-const functions.
+        if self.mode == Mode::Fn {
+            if let Lvalue::Temp(index) = *dest {
+                if self.temp_promotion_state[index as usize].is_promotable() {
+                    store(&mut self.temp_qualif[index as usize]);
+                }
+            }
+            return;
+        }
+
+        match *dest {
+            Lvalue::Temp(index) => store(&mut self.temp_qualif[index as usize]),
+            Lvalue::ReturnPointer => store(&mut self.return_qualif),
+
+            Lvalue::Projection(box Projection {
+                base: Lvalue::Temp(index),
+                elem: ProjectionElem::Deref
+            }) if self.mir.temp_decls[index as usize].ty.is_unique()
+               && self.temp_qualif[index as usize].map_or(false, |qualif| {
+                    qualif.intersects(Qualif::NOT_CONST)
+               }) => {
+                // Part of `box expr`, we should've errored
+                // already for the Box allocation Rvalue.
+            }
+
+            // This must be an explicit assignment.
+            _ => {
+                // Catch more errors in the destination.
+                self.visit_lvalue(dest, LvalueContext::Store);
+                self.statement_like();
+            }
+        }
+    }
+
+    /// Qualify a whole const, static initializer or const fn.
+    fn qualify_const(&mut self) -> Qualif {
+        let tcx = self.tcx;
+        let mir = self.mir;
+
+        let mut seen_blocks = BitVector::new(mir.basic_blocks.len());
+        let mut bb = START_BLOCK;
+        loop {
+            seen_blocks.insert(bb.index());
+
+            self.visit_basic_block_data(bb, &mir[bb]);
+
+            let target = match mir[bb].terminator().kind {
+                TerminatorKind::Goto { target } |
+                // Drops are considered noops.
+                TerminatorKind::Drop { target, .. } |
+                TerminatorKind::Call { destination: Some((_, target)), .. } => {
+                    Some(target)
+                }
+
+                // Non-terminating calls cannot produce any value.
+                TerminatorKind::Call { destination: None, .. } => {
+                    return Qualif::empty();
+                }
+
+                TerminatorKind::If { targets: (if_true, if_else), .. } => {
+                    // Need to allow bounds checking branches.
```

Comment:
Eurgh. This suggests to me that we may want to have special terminators for bounds-checks. (We could lower them to calls later on.)

I feel like this might make sense for other reasons -- e.g., the current MIR builder code code builds a lot of random tuples that seem expensive to build for normal cases, iirc.